app/src/main/java/com/github/se/icebreakrr/model/profile/ProfilesViewModel.kt

```diff
@@ -0,0 +1,148 @@
+   * @param gender Optional filter for gender.
+   * @param ageRange Optional filter for age range.
+   * @param tags Optional filter for specific tags.
+   */
+  fun getFilteredProfilesInRadius(
+      center: GeoPoint,
+      radiusInMeters: Double,
+      gender: Gender? = null,
+      ageRange: IntRange? = null,
+      tags: List<String>? = null
+  ) {
+    _loading.value = true
+    repository.getProfilesInRadius(
+        center = center,
+        radiusInMeters = radiusInMeters,
+        onSuccess = { profileList ->
+          val filteredProfiles =
+              profileList.filter { profile ->
+
+                // Filter by gender if specified
+                (gender == null || profile.gender == gender) &&
+
+                    // Filter by age range if specified
+                    (ageRange == null || profile.calculateAge() in ageRange) &&
+
+                    // Filter by tags if specified
+                    (tags == null || profile.tags.containsAll(tags))
+              }
+          _profiles.value = filteredProfiles
+          _loading.value = false
+        },
+        onFailure = { exception ->
+          _error.value = exception
+          _loading.value = false
+        })
+  }
+
+  /**
+   * Adds a new profile to the repository.
+   *
+   * @param profile The profile to be added.
+   */
+  fun addNewProfile(profile: Profile) {
+    _loading.value = true
+    repository.addNewProfile(
+        profile,
+        onSuccess = { _loading.value = false },
+        onFailure = { exception ->
+          _error.value = exception
+          _loading.value = false
+        })
+  }
+
+  /**
+   * Updates an existing profile in the repository.
+   *
+   * @param profile The profile with updated information.
+   */
+  fun updateProfile(profile: Profile) {
+    _loading.value = true
+    repository.updateProfile(
+        profile,
+        onSuccess = { _loading.value = false },
+        onFailure = { exception ->
+          _error.value = exception
+          _loading.value = false
+        })
+  }
+
+  /**
+   * Fetches a profile by its user ID (UID).
+   *
+   * @param uid The unique ID of the user whose profile is being retrieved.
+   */
+  fun getProfileByUid(uid: String) {
+    _loading.value = true
+    repository.getProfileByUid(
+        uid,
+        onSuccess = { profile ->
+          _selectedProfile.value = profile
+          _loading.value = false
+        },
+        onFailure = { exception ->
+          _error.value = exception
+          _loading.value = false
+        })
+  }
+
+  /**
+   * Deletes a profile by its user ID (UID).
+   *
+   * @param uid The unique ID of the user whose profile is to be deleted.
+   */
+  fun deleteProfileByUid(uid: String) {
+    _loading.value = true
+    repository.deleteProfileByUid(
+        uid,
+        onSuccess = { _loading.value = false },
+        onFailure = { exception ->
+          _error.value = exception
+          _loading.value = false
+        })
```

Comment:
Be careful about code duplication. onSucess and onFailure are always the same and use class variables. You could make two independant functions handleSuccess and handleFailure, and use them for each of your functions

Reply:
I did an handleError function to improve code reusability.